File: music-service/src/main/java/com/muco/musicservice/domain/interfaces/dto/ResponseDTO.java
package com.muco.musicservice.domain.interfaces.dto;

import com.fasterxml.jackson.annotation.JsonInclude;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record ResponseDTO(
        String message,
        Object data
) {
    private ResponseDTO(String message) {
        this(message, null);
    }

    public static ResponseDTO of(String message) {
        return new ResponseDTO(message);
    }

    public static <T> ResponseDTO of(String message, Object data, Class<T> clazz) {
        return new ResponseDTO(message, clazz.cast(data));
    }
}

File: chat-service/src/main/kotlin/com/muco/chatservice/domain/interfaces/dto/ResponseDTO.kt
package com.muco.chatservice.domain.interfaces.dto

import com.fasterxml.jackson.annotation.JsonInclude
import reactor.core.publisher.Mono

@JsonInclude(value = JsonInclude.Include.NON_NULL)
data class ResponseDTO(
    val message: String?,
    val data: Any?
) {

    private constructor(message: String): this(message, null)

    companion object {
        fun of(message: String): Mono<ResponseDTO> = Mono.just(ResponseDTO(message))
        fun of(message: String, data: Any?): Mono<ResponseDTO> = Mono.just(ResponseDTO(message, data ?: RuntimeException("데이터가 null 일 수 없습니다.")))
    }
}
